isis/tests/FunctionalTestsCsminit.cpp

```diff
@@ -0,0 +1,342 @@
+  ASSERT_EQ(infoGroup["ModelParameterTypes"].size(), 2);
+  EXPECT_EQ(infoGroup["ModelParameterTypes"][0].toStdString(), "FICTITIOUS");
+  EXPECT_EQ(infoGroup["ModelParameterTypes"][1].toStdString(), "REAL");
+
+  // Check the Kernels group
+  ASSERT_TRUE(testCube->hasGroup("Kernels"));
+  PvlGroup &kernGroup = testCube->group("Kernels");
+  EXPECT_TRUE(kernGroup.hasKeyword("ShapeModel"));
+}
+
+TEST_F(CSMPluginFixture, CSMinitRunTwice) {
+  // Run csminit twice in a row. Verify that end result is as if csminit were
+  // only run once with the second arguments.
+  QVector<QString> args = {
+    "from="+filename,
+    "isd="+isdPath};
+
+  UserInterface options(APP_XML, args);
+
+  csminit(options);
+
+  QVector<QString> altArgs = {
+    "from="+filename,
+    "isd="+altIsdPath,
+    "modelName="+QString::fromStdString(AlternativeTestCsmModel::SENSOR_MODEL_NAME)};
+
+  UserInterface altOptions(APP_XML, altArgs);
+
+  csminit(altOptions);
+
+  testCube->open(filename);
+
+  StringBlob stateString("", "CSMState");
+  testCube->read(stateString);
+  PvlObject blobPvl = stateString.Label();
+
+  // Verify contents of the StringBlob's PVL label
+  EXPECT_EQ(stateString.Name().toStdString(), "CSMState");
+  EXPECT_EQ(stateString.Type().toStdString(), "String");
+
+  // Check blob label ModelName and Plugin Name
+  EXPECT_EQ(QString(blobPvl.findKeyword("PluginName")).toStdString(), plugin->getPluginName());
+  EXPECT_EQ(QString(blobPvl.findKeyword("ModelName")).toStdString(), AlternativeTestCsmModel::SENSOR_MODEL_NAME);
+
+  // Check that the plugin can create a model from the state string
+  std::string modelName = QString(blobPvl.findKeyword("ModelName")).toStdString();
+  EXPECT_TRUE(plugin->canModelBeConstructedFromState(modelName, stateString.string()));
+
+  // Make sure there is only one CSMState Blob on the label
+  PvlObject *label = testCube->label();
+  label->deleteObject("String");
+  EXPECT_FALSE(label->hasObject("String"));
+
+  // Check that there is only one CsmInfo group
+  ASSERT_TRUE(testCube->hasGroup("CsmInfo"));
+  testCube->deleteGroup("CsmInfo");
+  ASSERT_FALSE(testCube->hasGroup("CsmInfo"));
+
+  // Check that there is only one ShapeModel
+  ASSERT_TRUE(testCube->hasGroup("Kernels"));
+  PvlGroup &kernGroup = testCube->group("Kernels");
+  EXPECT_TRUE(kernGroup.hasKeyword("ShapeModel"));
+  kernGroup.deleteKeyword("ShapeModel");
+  EXPECT_FALSE(kernGroup.hasKeyword("ShapeModel"));
+}
+
+TEST_F(CSMPluginFixture, CSMinitMultiplePossibleModels) {
+  // Test csminit when multiple possible models can be created. First, verify that this will fail
+  // without specifying the MODELNAME, then specify the MODELNAME and check the results.
+
+  QVector<QString> args = {
+    "from="+filename,
+    "isd="+altIsdPath};
+
+  UserInterface options(APP_XML, args);
+
+  // If there are two possible models, csminit will fail and prompt the user to specify the model
+  // and/or plugin name.
+  EXPECT_ANY_THROW(csminit(options));
```

Comment:
We used a lot of "hassubstr" calls during the testing sprint when we were expecting a specific error. e.g.:

EXPECT_THAT(Explain(ref_matcher, derived),HasSubstr("which cannot be dynamic_cast"));

Is that worth doing here?

Reply:
Yeah, this was a good idea!